rcl/test/rcl/test_subscription.cpp

```diff
@@ -885,6 +886,431 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_bad_return_l
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  rcl_subscription_t subscription = rcl_get_zero_initialized_subscription();
+  // not to set filter expression
+  rcl_subscription_options_t subscription_options = rcl_subscription_get_default_options();
+  ret = rcl_subscription_init(&subscription, this->node_ptr, ts, topic, &subscription_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_subscription_fini(&subscription, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  ASSERT_FALSE(rcl_subscription_is_cft_enabled(&subscription));
+
+  // failed to get filter
+  {
+    rcl_subscription_content_filter_options_t content_filter_options =
+      rcl_get_zero_initialized_subscription_content_filter_options();
+
+    ret = rcl_subscription_get_content_filter(
+      &subscription, &content_filter_options);
+    ASSERT_NE(RCL_RET_OK, ret);
+  }
+
+  ASSERT_TRUE(wait_for_established_subscription(&publisher, 30, 100));
+
+  // publish with a non-filtered data
+  constexpr char test_string[] = "NotFilteredData";
+  {
+    test_msgs__msg__Strings msg;
+    test_msgs__msg__Strings__init(&msg);
+    ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg.string_value, test_string));
+    ret = rcl_publish(&publisher, &msg, nullptr);
+    test_msgs__msg__Strings__fini(&msg);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  }
+
+  ASSERT_TRUE(wait_for_subscription_to_be_ready(&subscription, context_ptr, 30, 100));
+
+  {
+    test_msgs__msg__Strings msg;
+    test_msgs__msg__Strings__init(&msg);
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      test_msgs__msg__Strings__fini(&msg);
+    });
+    ret = rcl_take(&subscription, &msg, nullptr, nullptr);
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ASSERT_EQ(
+      std::string(test_string),
+      std::string(msg.string_value.data, msg.string_value.size));
+  }
+
+  // set filter
+  const char * filter_expression2 = "string_value = %0";
+  const char * expression_parameters2[] = {"'FilteredData'"};
+  size_t expression_parameters2_count = sizeof(expression_parameters2) / sizeof(char *);
+  bool is_cft_support{true};
+  {
+    rcl_subscription_content_filter_options_t options =
+      rcl_get_zero_initialized_subscription_content_filter_options();
+
+    EXPECT_EQ(
+      RCL_RET_OK,
+      rcl_subscription_content_filter_options_init(
+        filter_expression2, expression_parameters2_count, expression_parameters2,
+        &options)
+    );
+
+    ret = rcl_subscription_set_content_filter(
+      &subscription, &options);
+    if (RCL_RET_UNSUPPORTED == ret) {
+      is_cft_support = false;
```

Comment:
currently this fails with fastrtps but it's successful with connextdds, right?

Reply:
No, both rmw_fastrtps_cpp and rmw_connextdds are supported.

Reply:
could we use the compile definition `RMW_IMPLEMENTATION` to make sure this doesn't fail for fastrtps/connextdds

Reply:
The test doesn't fail. What I mean `fails with rmw_cyclonedds_cpp` is RCL_RET_UNSUPPORTED. Using `colcon test` should not get a failure.

Reply:
Yes.
What I mean is to have an `ASSERT_EQ(ret, RCL_RET_OK)` for both fastrtps and connextdds.

Reply:
I see you have already done that in https://github.com/ros2/rcl/pull/894/commits/630113369c87433e0bfe5d0f18b8ec73c267f7c7.